finality-aleph/src/data_io/status_provider.rs

```diff
@@ -0,0 +1,397 @@
+use crate::data_io::proposal::{AlephProposal, AvailableKind, ProposalStatus};
+use crate::BlockHashNum;
+use log::{error, warn};
+use lru::LruCache;
+use sc_client_api::HeaderBackend;
+use sp_runtime::traits::{Block as BlockT, Header as HeaderT, NumberFor};
+use sp_runtime::{generic::BlockId, SaturatedConversion};
+use std::{marker::PhantomData, sync::Arc};
+
+const BLOCKS_CACHE_SIZE: usize = 2000;
+const PROPOSALS_CACHE_SIZE: usize = 2000;
+
+pub trait ProposalStatusProvider<B: BlockT> {
```

Comment:
nothing personal, but I really do not like this solution :stuck_out_tongue_closed_eyes: things like `get_finalized_at` or `get_parent_hash` are not directly related to providing status. I would cut this API into (three?) parts:
 - a class `ProposalStatusProvider` providing proposal status with all these methods with default implementation
 - a trait for listening (`on_block_imported`, `on_block_finalized`)
 - a trait for querying storage

The class would keep subcomponents implementing these traits.

Reply:
Well, now I see that the 'listening' is rather artificial. So probably we could omit that trait

Reply:
And indeed use trait instead of bare class, because of two provided implementations.

Besides, I have doubts whether we need both implementations at this point

Reply:
Fine, this will likely change after a discussion with @timorl hopefully to something you like more.